### PR TITLE
Convert Depth Anything

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

[Depth Anything](https://twitter.com/_akhaliq/status/1749284669936275463) came out, and it is compatible with our implementation of DPT. It leverages DINOv2 as backbone.

It does use a small a tweak in the decoder, where it sets `size` instead of `scale_factor` when interpolating.

Demo notebook:  https://colab.research.google.com/drive/1tHrdu4TY6f_oTXJbqPUn2DVoSKxbNa3O?usp=sharing.

### CODE REVIEW

File: src/transformers/models/dpt/configuration_dpt.py [hunk 1]
```diff
@@ -96,6 +96,8 @@ class DPTConfig(PretrainedConfig):
             Whether to use bias in the pre-activate residual units of the fusion blocks.
         add_projection (`bool`, *optional*, defaults to `False`):
             Whether to add a projection layer before the depth estimation head.
+        use_size (`bool`, *optional*, defaults to `False`):
+            Whether to use the size for interpolation, rather than scale factor, as in Depth Anything.
```
Comment (amyeroberts) [inline src/transformers/models/dpt/configuration_dpt.py:100]: Let's use a better name here. From the name, it's not clear what `size` is being used from, it also doesn't extend well. At the moment, we're assuming that if we're not using `size`, then we're using scale factor, but what if a new model uses a different param? 

I'd suggest a parameter where one can select `"size"` or `"scale_factor"` as options

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    # Head
+    rename_keys.append(("pretrained.norm.weight", "backbone.layernorm.weight"))
+    rename_keys.append(("pretrained.norm.bias", "backbone.layernorm.bias"))
+
+    # activation postprocessing (readout projections + resize blocks)
+    # Depth Anything does not use CLS token => readout_projects not required
+
+    #     rename_keys.append((f"pretrained.act_postprocess{i+1}.0.project.0.weight", f"neck.reassemble_stage.readout_projects.{i}.0.weight"))
+    #     rename_keys.append((f"pretrained.act_postprocess{i+1}.0.project.0.bias", f"neck.reassemble_stage.readout_projects.{i}.0.bias"))
+
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:107]: If this code isn't used, then let's delete it 

```suggestion
```

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            return pad_size_left, pad_size_right
+
+        def __call__(self, img):
+            pads = list(itertools.chain.from_iterable(self._get_pad(m) for m in img.shape[-2:][::-1]))
+            output = torch.nn.functional.pad(img, pads)
+            return output
+
+        def __repr__(self):
+            return self.__class__.__name__ + "()"
+
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:206]:
```suggestion

```

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    def make_depth_transform() -> transforms.Compose:
+        return transforms.Compose(
+            [
+                transforms.ToTensor(),
+                lambda x: 255.0 * x[:3],  # Discard alpha component and scale by 255
+                transforms.Normalize(
+                    mean=(123.675, 116.28, 103.53),
+                    std=(58.395, 57.12, 57.375),
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:214]: Do these models have their image processor configs uplaoded with these values for normalization?

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        rename_key(state_dict, src, dest)
+    # read in qkv matrices
+    read_in_q_k_v(state_dict, config)
+
+    # load HuggingFace model
+    model = DPTForDepthEstimation(config)
+    model.load_state_dict(state_dict)
+    model.eval()
+
+    # TODO use image processor
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:259]: This looks like it's now done?

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        keep_aspect_ratio=True,
+        do_rescale=True,
+        do_normalize=True,
+        image_mean=[0.485, 0.456, 0.406],
+        image_std=[0.229, 0.224, 0.225],
+    )
+
+    import requests
+    from PIL import Image
+
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:273]: Why import these here ?

File: src/transformers/models/dpt/convert_depth_anything_to_hf.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    image = Image.open(requests.get(url, stream=True).raw)
+
+    # predict depth
+    import torchvision.transforms as T
```
Comment (amyeroberts) [inline src/transformers/models/dpt/convert_depth_anything_to_hf.py:278]: Or this here ?

File: src/transformers/models/dpt/modeling_dpt.py [hunk 8]
```diff
@@ -1046,9 +1058,17 @@ def __init__(self, config):
             self.projection = nn.Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
 
         features = config.fusion_hidden_size
+
+        image_size = (
+            config.backbone_config.image_size
+            if config.backbone_config is not None and not config.is_hybrid
+            else config.image_size
+        )
+        modifier = {"size": (int(image_size), int(image_size))} if config.use_size else {"scale_factor": 2}
```
Comment (amyeroberts) [inline src/transformers/models/dpt/modeling_dpt.py:1067]: This isn't robust as sometimes `image_size` can be a tuple in the model configs.

File: src/transformers/models/dpt/modeling_dpt.py [hunk 9]
```diff
@@ -1046,9 +1058,17 @@ def __init__(self, config):
             self.projection = nn.Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
 
         features = config.fusion_hidden_size
+
+        image_size = (
+            config.backbone_config.image_size
```
Comment (amyeroberts) [inline src/transformers/models/dpt/modeling_dpt.py:1063]: Let's not rely on the backbone config being present. The config being there or not defines how the backbone is loaded, but we shouldn't use it to parameterise the model itself

File: src/transformers/models/dpt/modeling_dpt.py [hunk 10]
```diff
@@ -700,11 +702,18 @@ def forward(self, hidden_states):
...
         # looping from the last layer to the second
-        for hidden_state, layer in zip(hidden_states[1:], self.layers[1:]):
-            fused_hidden_state = layer(fused_hidden_state, hidden_state)
+        for idx, (hidden_state, layer) in enumerate(zip(hidden_states[1:], self.layers[1:])):
+            size = (
+                hidden_states[1:][idx + 1].shape[2:]
+                if self.config.use_size and idx != len(hidden_states[1:]) - 1
+                else None
+            )
+            fused_hidden_state = layer(fused_hidden_state, hidden_state, size=size)
```
Comment (amyeroberts) [inline src/transformers/models/dpt/modeling_dpt.py:716]: Why is the layer behaviour being controlled in the forward pass here? The layers and modules should be set up with their paramterizations in their inits i.e. the `DPTFusionLayer` should know how it's rescaling, rather than having to recalculate it here on every forward pass

**Review by amyeroberts — COMMENTED** [review]
Thanks for the work adding this! 

Along with the other DPT extensions, this makes the forward passes complex with lots of if/else statements which should be avoided as much as possible. I think with reworking to set up the layers to be more in line with the rest of the transformers patterns this should be remedied

File: src/transformers/models/dpt/modeling_dpt.py (hunk 10 above)
Comment (NielsRogge) [reply src/transformers/models/dpt/modeling_dpt.py:716]: This is because the `size` is dynamic, and can only be determined by the size of the feature maps, which is inferred when doing a forward pass (since the model supports varying image resolutions).

**Comment by NielsRogge** [comment]
Have opened a PR that may be more in line with the philosophy. Let me know which one you prefer :)

**Comment by NielsRogge** [comment]
Closing in favor of #28654
